wordclock/si4703.c

```diff
@@ -0,0 +1,576 @@
+      } else if (group_code == 4 && group_ab == 'A') {
+        struct tm ts;
+        // Group 4A, clock time
+        // printf("RDS process CT: %04x %04x %04x %04x\n", block1, block2, block3, block4);
+        uint32_t modified_julian_date = getBits(block2, block3, 1, 17);
+
+        int year_utc  = (modified_julian_date - 15078.2) / 365.25;
+        int month_utc = (modified_julian_date - 14956.1 -
+                      (int)(year_utc * 365.25)) / 30.6001;
+        int day_utc   = modified_julian_date - 14956 - (int)(year_utc * 365.25) -
+                      (int)(month_utc * 30.6001);
+        if (month_utc == 14 || month_utc == 15) {
+          year_utc += 1;
+          month_utc -= 12;
+        }
+        year_utc += 1900;
+        month_utc -= 1;
+
+        int hour_utc   = getBits(block3, block4, 12, 5);
+        int minute_utc = getBits(0, block4, 6, 6);
+
+        double local_offset = (getBits(0, block4, 5, 1) ? -1 : 1) *
+                              getBits(0, block4, 0, 5) / 2.0;
+
+        bool is_date_valid = hour_utc <= 23 && minute_utc <= 59 &&
+                            (int)(local_offset) <= 14.0;
+        if (is_date_valid) {
+          int local_offset_hour = local_offset;
+          int local_offset_min  = (local_offset - (int)(local_offset)) * 60.0;
+          printf(
+                    "clock time: %04d-%02d-%02dT%02d:%02d:00%s%02d:%02d\n",
+                    year_utc, month_utc, day_utc,
+                    hour_utc, minute_utc, local_offset > 0 ? "+" : "-",
+                    local_offset_hour, abs(local_offset_min));
+
+          // TODO: day_utc must be yearday instead of monthday
+          ts.tm_year = year_utc - 1900;
+          ts.tm_mon = month_utc - 1;
+          ts.tm_mday = day_utc;
+          ts.tm_hour = hour_utc;
+          ts.tm_min = minute_utc;
+          ts.tm_sec = 0;
+          // struct timezone tzs = {0, 0};
+          // tzs.tz_minuteswest = -local_offset_hour - local_offset_min;
+          
+          // sntp_set_timezone(&tzs);
+
+          // This is UTC. Timezone and daylight saving time is done via the clock configuration page
+          time_t epoch = mktime(&ts);  
+          // time_t epoch = mktime(&ts) + local_offset_hour * 3600; 
+            // 0 + minute_utc*60 + hour_utc*3600 + day_utc*86400 +
+            // (year_utc-70)*31536000 + ((year_utc-69)/4)*86400 -
+            // ((year_utc-1)/100)*86400 + ((year_utc+299)/400)*86400;
+          // printf("RDS TIME RECEIVED, epoch: %u\n", (uint32_t)epoch);
+          return epoch;
+       }
+      }
+    }
+    SleepNI(41);
+  }
+  return rds_received; 
+}
+
+
+void si4703_task(void* p) {
+  int idx = 0;
+  const int RDS_TIMEOUT = 1000;
+  const int MAX_NO_TIME_COUNT = 130000/ RDS_TIMEOUT;
+  int no_time_count = 0;
+  SleepNI(2000);
+  while (!si4703_init()) {
+    si4703_dump_regs();
+    printf("si4703_init failed, disable si4703; no RDS clock time available\n");
+    si4703_powerOff();
+    SleepNI(2000);
```

Comment:
Alle Klocken die geen RDS hebben krijgen nu elke 2 seconde deze log regel.
Voor de hierbenik task heb ik toen ooit een conditionele task start gemaakt is dat hier ook niet een mooie oplossing?